### DISCUSSION TITLE
Help with creating new document

### DISCUSSION THREAD

**Discussion by Rol3oT** [GodotNuts/GodotFirebase#449] [Q&A] [answered]
New to exploring Firebase, have been following this tutorial and modifying code to fit Godot 4.3 
https://youtu.be/7ehE4IdSP5s?t=853  @14:00 
In the tutorial person uses:
`collection.update(doc_id, data)`
and from his explanation if document doesn't exist, GodotFirebase will create one! Great!

Looking at the documentation for GodotFirebase for Godot 4.x :
`4.x:
await collection.update(document : FirestoreDocument) -> FirestoreDocument # Updated document is returned`

so I have to pass in document as an argument, I have been looking at this a little and I am missing on how to create document object, looks like I can receive it with the `collection.get` function...

I would prefer to use `update` function since it will create document if there is no document and update values of the existing document(Sounds handy!)

If thats not possible I understand how `collection.add` function works. However having trouble understanding how `collection.update` function supposed to be used, wish examples were better in documentation. 

Could anyone throw in some examples of `collection.update` please?

**Comment by WolfgangSenff** [comment]
First off, you should not really be following that tutorial - it's wrong because it's out of date. That said, others have been asking about this recently, so I'll probably try to add something back in to allow it. It really shouldn't work and the way the code in the addon is now is intentional, but enough people want it back. I'll look into it. It might take several days though, have a ton going on right now!

**Reply by WolfgangSenff** [reply, marked as answer]
Oh, and ill try to take some time to update the documentation. You can check out the test harness which should be up to date here: https://github.com/GodotNuts/GodotFirebaseTestHarness

Look in the tests and look for the ones for Firestore. It does an update, should be pretty clear (make sure you switch to the 4.x branch first, as I haven't updated it to main yet in the test harness like I have in the addon).

**Reply by Rol3oT** [reply]
Thank you for your reply! I am mostly using that tutorial as a template, something to start with. I was able successfully modify code and get authentication working.

Thank you for sharing that test harness code! What I understand is:
You make your dictionary, collection and then you can create document on Firebase using `add` function

```
var DefaultDocument = { 'name': 'Document1', 'points': 20, 'crud': { 'something': 'other thing', 'main': { 'info' : 8 }}, 'active': 'true', "server_timestamp_attempt": null, "increment_field": 0, "decrement_field": 0, "max_field": 5, "min_field": 2 }

_collection = Firebase.Firestore.collection('Firebasetester')
_document = await _collection.add("Document1", DefaultDocument)

```

But for updating existing document do I have to call `get_doc` in order to have a FirestoreDocument instance?
```
_document = await _collection.get_doc('Document1')
var key = "name"
_document.remove_field(key)
	
_document = await _collection.update(_document)
```

Can I not create a FirestoreDocument instance without calling `add` or `get` to safe on the database read tokens?
Like
```
var _document = new FirestoreDocument(DefaultDocument)
_document.remove_field(key)
	
_document = await _collection.update(_document)
```


It seems like from _FirestoreDocument_ class it is possible
```
func _init(doc : Dictionary = {}):
	_transforms = FieldTransformArray.new()
	
	if doc.has("fields"):
		document = doc.fields
	if doc.has("name"):
		doc_name = doc.name
		if doc_name.count("/") > 2:
			doc_name = (doc_name.split("/") as Array).back()
	if doc.has("createTime"):	
		self.create_time = doc.createTime
```

but I remember I had trouble with that last night. Am I doing it right?

**Reply by WolfgangSenff** [reply]
Right now, I don't believe you can do that, re: add and update all at once. That's what I would be adding back in functionally. I have that code in there to support it in the future, but I'm not actually sure it works. In fact I think it does not, but I'm not certain.
